docs/compute-engine/model_optimization_guide.md

```diff
@@ -0,0 +1,108 @@
+    the ReLU before the batch-norm for a sequence of binary convolutions: if the
+    output of a ReLU were fed directly into a subsequent binary convolution, the
+    input would be quantised to constant ones and all information would be
+    destroyed. As a result, the LCE converter supports op fusing of the layer
+    sequence `QuantConv2D` → `ReLU` → `BatchNormalization`.
+
+    Note that a batch-norm can still be fused into a `QuantConv2D` layer without
+    a ReLU between them; however, a subsequent ReLU would not also be fused.
+
+*   **Avoid using Larq `QuantConv2D` layers with 'same-zero' padding**
+
+    The [Keras `Conv2D`
+    layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv2D)
+    supports either "valid" padding (no input padding) or "same" padding
+    (padding the input with zeroes so that the output has the same spatial
+    resolution). The Larq [`QuantConv2D`](/larq/api/layers/#quantconv2d) layer
+    also supports "valid" and "same" padding, but additionally has an attribute
+    `pad_values` that controls the padding constant that is used in the "same"
+    padding case. We refer to the combination of `padding="same"` with
+    `pad_values=0.0` (the default, to match the `Conv2D` layer) as 'same-zero'
+    padding, and the combination with `pad_values=1.0` as 'same-one' padding.
+
+    It is strongly recommended to use 'same-one' padding instead of 'same-zero'
+    padding; for an example of models using 'same-one' padding, see the
+    [QuickNet models in Larq
+    Zoo](https://github.com/larq/zoo/blob/bb0dbf7c13bcb21149cd19c3bd51fe7d885b3bd8/larq_zoo/sota/quicknet.py#L107-L108).
+
+    With 'same-zero' padding, because zero cannot be naturally represented in
+    binary data (which uses a single bit to represent either +1 or -1) a
+    zero-padding correction has to be added to values at the edge of the output
+    tensor in order to give the correct result. This extra computation can be
+    avoided by using 'same-one' padding.
+
+    Additionally, 'same-zero' padding prevents the converter op fusing
+    optimisations described above. These optimisations are not affected by
+    'same-one' padding.
+
+*   **Replace `QuantDense` layers with 1×1 `QuantConv2D`s**
+
+    LCE does not yet include an optimised binary dense op, which means that Larq
+    `QuantDense` layers are not accelerated in any way. The recommended
+    workaround for this is to replace all `QuantDense` layers with an equivalent
+    sequence of reshapes and a 1×1 `QuantConv2D` layer:
+
+    ```python
+    # Replace this...
+    x = lq.layers.QuantDense(num_dense_outputs, ...)(x)
+
+    # ...with this
+    x = tf.keras.layers.Reshape((1, 1, -1))(x)
+    x = lq.layers.QuantConv2D(num_dense_outputs, kernel_size=(1, 1), ...)(x)
+    x = tf.keras.layers.Reshape((-1,))(x)
+    ```
+
+*   **Ensure that `filter_height` × `filter_width` × `input_channels` is a
+    multiple of 128**
```

Comment:
For the common filter sizes `1x1`, `3x3` and `5x5` (or any filter of which the number of pixels isn't divisible by two), to only way to achieve this constraint is to use multiples of 128 input channels, which is quite a strong constraint I think. Therefore I'm a bit unsure about how this part should be phrased. Maybe we can say that this optimization is less important than the one about multiples of 32 channels. (The multiple of 32 affects bitpacking, im2col, ruy packing and bgemm, whereas this particular constraint only affects ruy packing and bgemm)

Reply:
Yeah agreed, I'll try and think about how to reword this

Reply:
@koenhelwegen Have you observed this effect in your benchmarks? i.e. a `1x1` convolution on 64 input channels should be only slightly faster than on 128 input channels instead of 2X faster (bitpacking and im2col should be 2X faster, whereas the main bgemm part should be equally fast).

Reply:
The script currently only allows same input and output channels, but it seems to scale pretty quadratically in channels, so presumably linear in input channels (width/height = 8, kernel_size=1, bitpacked in/out = True):

| in/out channels | block latency (ms) | im2col (ms) | bgemv (ms) |
| --- | --- | --- | --- |
| 64 | 2.28 | 0.0885 | 2.195125 |
| 128 | 8.07 | 0.107 | 7.932125 |

Reply:
Without this constraint, we'd expect the 64 case to be 4X faster than the 128 case, so it would be about 2.0 ms, but due to this padding effect we expect it to be anywhere between 2.0 and 8.0 ms. It appears that its about 2.3 ms so apparently this effect is very small. I would have expected a bigger impact to be honest, but I guess this is better. Maybe we can say that we measured the impact of this and that it definitely is measurable but it's not very strong?

Reply:
Yes this seems to be true on pixel1. For a 1x1 kernel we get the following:

| width/height | in/out channels | latency (ms) |
| --- | --- | --- |
| 8 | 32 | 0.008 |
| 8 | 128 | 0.020 |
| 32 | 32 | 0.103 |
| 32 | 128 | 0.298 |

A naive estimate would indicate the 128 channel blocks should be 16x slower, but it's less than a factor 4 in both cases.

Reply:
The effect seems to be pretty strong then. However usually binary convs are `3x3` in which case the `9 * 32` gets rounded up to `12 * 32` (33% extra for 32 input channels) and `9 * 64` gets rounded to `10 * 64` (11% extra for 64 input channels).

So for `1x1` kernels the effect is strong (2X for 64 channels, 4X for 32 channels), but for `3x3` kernels the effect would be 11% instead of 2X and 33% instead of 4X.